cli/factory.rs

```diff
@@ -1052,3 +1066,230 @@ impl CliFactory {
     })
   }
 }
+
+#[derive(Debug, Copy, Clone)]
+pub struct SpecifierInfo {
+  /// Type check as an ES module.
+  pub check: bool,
+  /// Type check virtual modules from doc snippets. If this is set but `check`
+  /// is not, this may be a markdown file for example.
+  pub check_doc: bool,
+}
+
+struct WorkspaceFileContainerEntry {
+  specifiers: Vec<(ModuleSpecifier, SpecifierInfo)>,
+  doc_snippet_specifiers: Vec<ModuleSpecifier>,
+  factory: CliFactory,
+  worker_factory: Arc<CliMainWorkerFactory>,
+}
+
+impl WorkspaceFileContainerEntry {
+  fn checked_specifiers(&self) -> impl Iterator<Item = &ModuleSpecifier> {
+    self
+      .specifiers
+      .iter()
+      .filter_map(|(s, i)| i.check.then_some(s))
+      .chain(self.doc_snippet_specifiers.iter())
+  }
+}
+
+pub struct WorkspaceFileContainer {
+  entries: Vec<WorkspaceFileContainerEntry>,
+}
+
+impl WorkspaceFileContainer {
+  #[allow(clippy::type_complexity)]
+  pub async fn from_workspace_dirs_with_files<T: Clone>(
+    mut workspace_dirs_with_files: Vec<(Arc<WorkspaceDirectory>, FilePatterns)>,
+    factory: &CliFactory,
+    extract_doc_files: Option<fn(File) -> Result<Vec<File>, AnyError>>,
+    collect_specifiers: fn(
+      FilePatterns,
+      Arc<CliOptions>,
+      Arc<FileFetcher>,
+      T,
+    ) -> std::pin::Pin<
+      Box<
+        dyn Future<
+          Output = Result<Vec<(ModuleSpecifier, SpecifierInfo)>, AnyError>,
+        >,
+      >,
+    >,
+    args: T,
+  ) -> Result<Self, AnyError> {
+    workspace_dirs_with_files.sort_by_cached_key(|(d, _)| d.dir_url().clone());
+    let cli_options = factory.cli_options()?;
+    let watcher_communicator = &factory.watcher_communicator;
+    let all_scopes = Arc::new(
+      workspace_dirs_with_files
+        .iter()
+        .filter(|(d, _)| d.has_deno_or_pkg_json())
+        .map(|(d, _)| d.dir_url().clone())
+        .collect::<BTreeSet<_>>(),
+    );
+    let dir_count = workspace_dirs_with_files.len();
+    let mut entries = Vec::with_capacity(dir_count);
+    for (workspace_dir, files) in workspace_dirs_with_files {
+      let (npmrc, _) = discover_npmrc_from_workspace(&workspace_dir.workspace)?;
+      let lockfile =
+        CliLockfile::discover(&cli_options.flags, &workspace_dir.workspace)?;
+      let scope_options = (dir_count > 1).then(|| ScopeOptions {
+        scope: workspace_dir
+          .has_deno_or_pkg_json()
+          .then(|| workspace_dir.dir_url().clone()),
+        all_scopes: all_scopes.clone(),
+      });
+      let cli_options = Arc::new(CliOptions::new(
+        cli_options.flags.clone(),
+        cli_options.initial_cwd().to_path_buf(),
+        lockfile.map(Arc::new),
+        npmrc,
+        workspace_dir,
+        false,
+        scope_options.map(Arc::new),
+      )?);
+      let mut factory = CliFactory::from_cli_options(cli_options.clone());
```

Comment:
Ideally we create a single factory and then from that single factory we create services. Right now this seems to be doing a lot of setup code over and over for each workspace directory? Is there a way we could change this not to do that? Why is this necessary?

Reply:
Also responding to https://github.com/denoland/deno/pull/27009#pullrequestreview-2493598738.

It's an implementation detail of this struct to avoid doing a larger refactor for now. The services in `CliFactory` are based on the config from the workspace directory in `cli_options.start_dir`, and deeply attached to `cli_options.start_dir.workspace()`.

To expound on that, I want this to be initial infrastructure for cross-config support without workspaces (i.e. across multiple workspaces). That means it needs to support completely different resolution, vendor/cache locations and other graph-building settings across different workspace dirs. The LSP already supports this for reference.

Currently the `resolve_foo_options_for_members()` methods don't support cross-workspace, but the point is to not further entrench that limitation.

Reply:
I don't think we should support more than one workspace in most of the CLI code. I think that would lead to too much complexity and is a fair limitation to have in the CLI. Let's remove that for now and focus on just having this work within one workspace.

Reply:
In the future, if we add support for multiple workspaces then we should do a larger refactor that specifically does that, but I think that's too much to add for this PR.

Reply:
> I think we should remove that for now and focus on just having this work within one workspace.

I can plumb the compiler options through the check invocation rather than keeping it in the state, which would allow sharing the root factory for now I think. I'll see if everything still works with that.

> I don't think we should support more than one workspace in most of the CLI code. I think that would lead to too much complexity and is a fair limitation to have in the CLI.

Strongly disagree, it's not fair that if users want their subproject configurations to be differentially respected, they need to use workspaces. To the contrary they should use workspaces when they have shared configuration to factor out. There's no more complexity in splitting args, invoking many times and optimising for shared state than we already have to do with single-workspace-many-members. It's just complex to refactor what we have now.

Reply:
If we do do that, then we could stick to keeping it all behind `Workspace` without changing anything in the CLI. We already support multiple workspaces there with the `patch` feature.

Reply:
> I can plumb the compiler options through the check invocation rather than keeping it in the state, which would allow sharing the root factory for now I think. I'll see if everything still works with that.

Maybe in `TypeChecker` we could see the current `ModuleGraph` that gets passed in and from that segment the graph based on all the different kinds of compiler options found from the `CliOptions`. Then we wouldn't need to plumb the compiler options through I think.

Reply:
Okay I was reminded why this is an issue even in single-workspace, there is no shared `ModuleGraph` because in the very least `JsxImportSourceConfig` will vary with compiler options and is a resolver parameter.

We could only respect the one specified in the compiler options of the 'default' `cli_options.start_dir` (btw I hate this, there shouldn't be such a thing), but IIRC this was stressed as a "do it right or not at all" point during a meeting.

We could give the resolver `BTreeMap<Url, JsxImportSourceConfig>`, but we'd _really_ be digging ourselves into the single-workspace assumption for this to affect `deno_graph`'s API

It would be simplest and most correct to just handle this like editors do. If two files from different configuration scopes are specified at the CLI (shared workspace or not), assume little about what settings they share and construct different graph-building/cache services for each.

I'm looking over the code base and `CliFactory` is constructed relatively late in the process for each command. It's not even shared between watched runs for instance. Other than maybe its name there's nothing strange about having one-per-workspace-dir.

Reply:
We need any solution that doesn't create multiple factories and a solution that creates one graph. Introducing multiple factories into the CLI is too much complexity and will lead to bugs.

>  but we'd really be digging ourselves into the single-workspace assumption for this to affect deno_graph's API

A `Workspace` can be changed to support any number of deno workspaces. It already has a bit of support for this with the patch feature.

> Okay I was reminded why this is an issue even in single-workspace, there is no shared ModuleGraph because in the very least JsxImportSourceConfig will vary with compiler options and is a resolver parameter.

Let's update deno_graph to ask based on the referrer. We need this anyway for `deno run`.

Let's get this fix landed with minimal foundational design changes and without creating multiple factories. Let's consider them out of scope for this PR.